src/core/net/nat64_translator.cpp

```diff
@@ -0,0 +1,474 @@
+        LogWarn("incoming message is an IPv4 datagram but no IPv4 CIDR for NAT64 configured, drop");
+        ExitNow(res = kForward);
+    }
+
+    if (!mNat64Prefix.IsValidNat64())
+    {
+        LogWarn("incoming message is an IPv4 datagram but no NAT64 prefix configured, drop");
+        ExitNow(res = kDrop);
+    }
+
+    if (ip4Header.GetTtl() <= 1)
+    {
+        LogDebg("incoming datagram TTL reached");
+        ExitNow(res = kDrop);
+    }
+    ip4Header.SetTtl(ip4Header.GetTtl() - 1);
+
+    mapping = FindMapping(ip4Header.GetDestination());
+    if (mapping == nullptr)
+    {
+        LogWarn("no mapping found for the IPv4 address");
+        ExitNow(res = kDrop);
+    }
+
+    aMessage.RemoveHeader(sizeof(Ip4::Header));
+
+    ip6Header.Clear();
+    ip6Header.InitVersionTrafficClassFlow();
+    ip6Header.GetSource().SynthesizeFromIp4Address(mNat64Prefix, ip4Header.GetSource());
+    ip6Header.SetDestination(mapping->mIp6);
+    ip6Header.SetFlow(0);
+    ip6Header.SetHopLimit(ip4Header.GetTtl());
+
+    // Note: TCP and UDP are the same for both IPv4 and IPv6 except for the checksum calculation, we will update the
+    // checksum in the payload later. However, we need to translate ICMPv6 messages to ICMP messages in IPv4.
+    switch (ip4Header.GetProtocol())
+    {
+    case Ip4::kProtoUdp:
+        ip6Header.SetNextHeader(Ip6::kProtoUdp);
+        res = kForward;
+        break;
+    case Ip4::kProtoTcp:
+        ip6Header.SetNextHeader(Ip6::kProtoTcp);
+        res = kForward;
+        break;
+    case Ip4::kProtoIcmp:
+        ip6Header.SetNextHeader(Ip6::kProtoIcmp6);
+        SuccessOrExit(TranslateIcmp4(aMessage));
+        res = kForward;
+        break;
+    default:
+        ExitNow(res = kDrop);
+    }
+
+    // res here must be kForward based on the swich above.
+    // TODO: Implement the logic for replying ICMP datagrams.
+    ip6Header.SetPayloadLength(aMessage.GetLength() - aMessage.GetOffset());
+    Checksum::UpdateMessageChecksum(aMessage, ip6Header.GetSource(), ip6Header.GetDestination(),
+                                    ip6Header.GetNextHeader());
+    if (aMessage.Prepend(ip6Header) != kErrorNone)
+    {
+        // This might happen when the platform failed to reserve enough space before the original IPv4 datagram.
+        LogWarn("failed to prepend IPv6 head to translated message");
+        ExitNow(res = kDrop);
+    }
+    aMessage.SetType(Message::kTypeIp6);
+
+exit:
+    return res;
+}
+
+Translator::AddressMapping::InfoString Translator::AddressMapping::ToString(void)
+{
+    InfoString string;
+
+    string.Append("%s -> %s", mIp6.ToString().AsCString(), mIp4.ToString().AsCString());
+
+    return string;
+}
+
+void Translator::ReleaseMapping(AddressMapping &aMapping)
+{
+    LogInfo("mapping removed: %s", aMapping.ToString().AsCString());
+    IgnoreError(mIp4AddressPool.PushBack(aMapping.mIp4));
+    mAddressMappingPool.Free(aMapping);
```

Comment:
nit: since "removed" is past-tense
```suggestion
    IgnoreError(mIp4AddressPool.PushBack(aMapping.mIp4));
    mAddressMappingPool.Free(aMapping);
    LogInfo("mapping removed: %s", aMapping.ToString().AsCString());
```